README.md
# ReactCustomTable

Render table in react, fully customizable, compatible with redux.

## Our features:
- We render table and handle pagination, the data passed into our table is only for the CURRENT PAGE rather than the entire collection of data, so in this case we didn't implement the data layer logic for sorting as well as the pagination, however, we provide the callbacks with all the params you need (sorted column, sorting direction, current page) to you, you can implement the logic yourself (usually for server side query).
- We are not including any css, you can use whatever css library (i.e. Bootstrap, Foundation) you want, and it is easy to integrate.
- This plugin is designed to work with redux. For all data query events (i.e. sorting, changing page), we provide a callback with the current state of the table (i.e. active page, sorted column, sorting direction), you can easily hook it up with your action (i.e. trigger ajax request) for redux to get the new data back to the table.
- Since the UI and DOM can be totally different for users, we didn't implement the additional features like search and page size selector. However, as long as you define your component within our scope (under TableContainer), you will get all the data you need in your component props. You can easily extend the functionality to do something like keywords search, delete selected rows, edit selected row, etc. In our demo, we will show you how to add those components by yourself.

## Demo

Live Demo: https://da-kuang.github.io/react-custom-table/

Run demo locally:
```
git clone https://github.com/Da-Kuang/react-custom-table.git
npm run demo
http://localhost:8080
```

## Install

```
npm install react-custom-table --save
```

## Usage

### To generate a simple table

```
import React from 'react';
import { TableContainer, Table } from 'react-custom-table';

<TableContainer
    columns={[
    	{id: "name", title: "Name"},
    	{id: "gender", title: "Gender"}
    ]}
    rows={[
    	{id: "1", name: "David", gender: "Male"},
        {id: "2", name: "Kelly", gender: "Female"},
        {id: "3", name: "James", gender: "Male"},
        {id: "4", name: "Tim", gender: "Male"}
    ]} >

    <Table/>

</TableContainer>
```

### To generate a table with pagination

```
import React from 'react';
import { TableContainer, Table, Paginator } from 'react-custom-table';

<TableContainer
    columns={[
    	{id: "name", title: "Name"},
    	{id: "gender", title: "Gender"}
    ]}
    rows={[
    	{id: "1", name: "David", gender: "Male"},
        {id: "2", name: "Kelly", gender: "Female"},
        {id: "3", name: "James", gender: "Male"},
        {id: "4", name: "Tim", gender: "Male"}
    ]}>

    <Table/>

    <Paginator total={7} pageSize={4} />
</TableContainer>
```

### A more complicated example

```
import React from 'react';
import { TableContainer, Table, Paginator } from 'react-custom-table';

const columns = [
    {id: "name", title: "Name", sortable: true},
    {id: "gender", title: "Gender", template: (v) => <b style={{color: v.gender === "Male" ? 'blue' : 'pink'}}>{v.gender}</b>},
    {id: "age", title: "Age", sortable: true},
    {id: "height", title: "Height", sortable: false}
]
const pageData = [
    {id: "1", name: "David", gender: "Male", age: "17", height: "176"},
    {id: "2", name: "Kelly", gender: "Female", age: "19", height: "180"},
    {id: "3", name: "James", gender: "Male", age: "12", height: "150"},
    {id: "4", name: "Tim", gender: "Male", age: "39", height: "180"}
]

<TableContainer
    containerClass="table-responsive"
    columns={columns}
    rows={pageData}
    onSort={this.handleSort}
    onPageChange={this.handlePageClick}>

    <Table
        tableClass="table table-bordered table-hover"
        sortableClass="sortable"
        sortDescClass="desc"
        sortAscClass="asc"
        selectedRowClass="selected-row"
        selectable={true}
        selectColumnTemplate={(v) => <input type='checkbox' checked={v}/>}
        selectAllTemplate={(v) => <input type='checkbox' checked={v}/>}/>

    <Paginator
        paginatorClass="pagination pull-right"
        total={7}
        firstPageTemplate={() => <a>{"<<"}</a>}
        prevPageTemplate={() => <a>{"<"}</a>}
        pageTemplate={(p) => <a>{p}</a>}
        nextPageTemplate={() => <a>{">"}</a>}
        lastPageTemplate={() => <a>{">>"}</a>}
        pageSize={4}/>

</TableContainer>
```

To see more details for the usage, please check the code in our demo folder.

CHANGES.md
1.0.0
==========
- A bunch of bug fixes
- Change some prop names which may not be back-compatible
- Pass all state of the table to any children component

0.0.3
==========
- update README
- npm update


0.0.2
==========
- initial version for react-custom-table
- sorting
- pagination
- row selection


src/Table.js
import React, { Component, PropTypes } from 'react'
import classNames from 'classnames/bind'


export default class Table extends Component {

  static propTypes = {
    idKey: PropTypes.string,
    tableClass: PropTypes.string,
    sortableClass: PropTypes.string,
    sortDescClass: PropTypes.string,
    sortAscClass: PropTypes.string,
    selectedRowClass: PropTypes.string,
    sortCol: PropTypes.string,
    sortDesc: PropTypes.bool,
    selectable: PropTypes.bool,
    selected: PropTypes.array,
    columns: PropTypes.array,
    rows: PropTypes.array,
    noResultTemplate: PropTypes.func,
    selectColumnTemplate: PropTypes.func,
    onSort: PropTypes.func,
    onRowClick: PropTypes.func,
    onSelectAll: PropTypes.func
  }

  static defaultProps = {
    idKey: 'id',
    selectable: false,
    columns: [],
    rows: [],
    noResultTemplate: (cols) => {
      return (
        <tbody>
          <tr>
            <td colSpan={cols} style={{textAlign: "center"}}>
              Sorry, we could not find any data!
            </td>
          </tr>
        </tbody>
      )
    }
  }

  handleSort(colId){
    const { onSort } = this.props

    onSort && onSort(colId)
  }

  handleRowClick(rowId){
    const { onRowClick } = this.props

    onRowClick && onRowClick(rowId)
  }

  renderTableHead(){
    const { idKey,
            rows,
            columns,
            selected,
            selectable,
            sortableClass,
            sortDescClass,
            sortAscClass,
            sortCol,
            sortDesc,
            selectColumnTemplate,
            selectAllTemplate,
            onSelectAll } = this.props

    const rowIds = rows.map(row => row[idKey])

    return (
      <thead>
        <tr>
          {selectColumnTemplate ?
            <td onClick={onSelectAll ? onSelectAll.bind(null, rowIds) : null}>
              {selectAllTemplate && selectAllTemplate(rowIds.length > 0 && rowIds.every(id => selected.indexOf(id) > -1))}
            </td>
            : null
          }

          {columns.map(col =>
            <th
              key={`th-${col.id}`}
              className={classNames({
                [sortableClass]: sortableClass && col.sortable,
                [sortDescClass]: sortDescClass && col.id === sortCol && sortDesc,
                [sortAscClass]: sortAscClass && col.id === sortCol && !sortDesc
              })}
              onClick={col.sortable ? this.handleSort.bind(this, col.id): null}>

              { col.title }

            </th>
          )}
          </tr>
      </thead>
    )
  }

  renderTableBody(){
    const { idKey,
            columns,
            rows,
            selectedRowClass,
            selected,
            selectColumnTemplate } = this.props

    return (
      <tbody>
        {rows.map(row =>
          <tr
            key={`tr-${row[idKey]}`}
            className={classNames({
              [selectedRowClass]: selectedRowClass && selected.indexOf(row[idKey]) > -1
            })}
            onClick={selectColumnTemplate ? this.handleRowClick.bind(this, row[idKey]) : null}>

              {selectColumnTemplate ? <td>{selectColumnTemplate(selected.indexOf(row[idKey]) > -1)}</td> : null}

              {columns.map(col =>
                <td key={`td-${row[idKey]}-${col.id}`}>
                  {col.template ? col.template(row) : row[col.id]}
                </td>
              )}

          </tr>
        )}
      </tbody>
    )
  }

  render(){
    const { tableClass,
            noResultTemplate,
            selectColumnTemplate,
            rows,
            columns } = this.props

    const tableColNumber = columns.length + (selectColumnTemplate ? 1 : 0)

    return (
      <table className={tableClass}>
        { this.renderTableHead() }
        { rows.length > 0 ? this.renderTableBody() : noResultTemplate(tableColNumber) }
      </table>
    )
  }
}

src/Paginator.js
import React, { Component, PropTypes } from 'react'


export default class Paginator extends Component {

  static propTypes = {
    paginatorClass: PropTypes.string,
    activePageClass: PropTypes.string,
    total: PropTypes.number,
    pageSize: PropTypes.number,
    activePage: PropTypes.number,
    pageLimit: PropTypes.number,
    pageTemplate: PropTypes.func,
    nextPageTemplate: PropTypes.func,
    prevPageTemplate: PropTypes.func,
    firstPageTemplate: PropTypes.func,
    lastPageTemplate: PropTypes.func,
    onPageClick: PropTypes.func
  }

  static defaultProps = {
    total: 0,
    pageSize: 10,
    activePage: 0,
    pageLimit: 5,
    activePageClass: 'active'
  }

  constructor(props){
    super(props)

    const { total, pageSize } = props

    this.totalPages = total % pageSize === 0 ? parseInt(total / pageSize) : parseInt(total / pageSize) + 1
  }

  componentWillUpdate(props){
    const { total,
            pageSize,
            activePage,
            onPageClick } = props

    this.totalPages = total % pageSize === 0 ? parseInt(total / pageSize) : parseInt(total / pageSize) + 1

    if (this.totalPages - 1 < activePage)
      onPageClick && onPageClick(0)
  }

  handlePageClick(page){
    const { onPageClick } = this.props

    onPageClick && onPageClick(page)
  }

  renderFirstPage(){
    const { firstPageTemplate,
            activePage } = this.props

    return (
      firstPageTemplate && activePage > 0 ?
        <li onClick={this.handlePageClick.bind(this, 0)}>
          {firstPageTemplate()}
        </li> : null
    )
  }

  renderPrevPage(){
    const { prevPageTemplate,
            activePage } = this.props

    return (
      prevPageTemplate && activePage > 0 ?
        <li onClick={this.handlePageClick.bind(this, activePage - 1)}>
          {prevPageTemplate()}
        </li> : null
    )
  }

  renderPages(){
    const { activePageClass,
            pageLimit,
            total,
            pageTemplate,
            activePage } = this.props

    let pagelowerBound = activePage - pageLimit / 2
    let pageUpperBound = activePage + pageLimit / 2

    if (pagelowerBound < 0){
      pagelowerBound = 0
      pageUpperBound = pageLimit - 1
    }

    if (pageUpperBound > total - 1){
      pagelowerBound = total - 1
      pageUpperBound = total - pageLimit
    }

    return (
       Array(this.totalPages).fill().map((_, i) =>
        {
          return (
            i >= pagelowerBound && i <= pageUpperBound ?
              <li
                key={i}
                onClick={this.handlePageClick.bind(this, i)}
                className={i === activePage ? activePageClass : null}>
                {pageTemplate ? pageTemplate(i) : i}
              </li> : null
          )
        })
    )
  }

  renderNextPage(){
    const { nextPageTemplate,
            activePage } = this.props

    return (
      nextPageTemplate && activePage < this.totalPages - 1 ?
        <li onClick={this.handlePageClick.bind(this, activePage + 1)}>
          {nextPageTemplate()}
        </li> : null
    )
  }

  renderLastPage(){
    const { lastPageTemplate,
            activePage } = this.props

    return (
      lastPageTemplate && activePage < this.totalPages - 1 ?
        <li onClick={this.handlePageClick.bind(this, this.totalPages - 1)}>
          {lastPageTemplate()}
        </li> : null
    )
  }

  render(){
    const { paginatorClass,
            total,
            pageSize,
            pageLimit,
            extraTemplate,
            activePage } = this.props

    return (
      <ul className={paginatorClass}>
        { this.renderFirstPage() }
        { this.renderPrevPage() }
        { this.renderPages() }
        { this.renderNextPage() }
        { this.renderLastPage() }
      </ul>
    )
  }
}

src/TableContainer.js
import React, { Component, cloneElement, Children, PropTypes }  from 'react'
import Table from './Table'
import Paginator from './Paginator'
import { Set } from 'immutable'


class TableContainer extends Component {

  static propTypes = {
    containerClass: PropTypes.string,
    columns: PropTypes.array,
    rows: PropTypes.array,
    onRowSelect: PropTypes.func,
    onAllRowsSelect: PropTypes.func,
    onPageChange: PropTypes.func,
    onSort: PropTypes.func
  }

  constructor(props){
    super(props)

    this.state = {
      sortCol: null,
      sortDesc: false,
      selectedRows: Set(),
      activePage: 0
    }
  }

  handlePageClick(page){
    const { onPageChange } = this.props

    const { sortCol, sortDesc } = this.state

    this.setState({
      activePage: page
    })

    onPageChange && onPageChange(page, sortCol, sortDesc)
  }

  handleSort(colId){
    const { sortCol, sortDesc, activePage } = this.state
    const { onSort } = this.props

    const newSortDesc = colId === sortCol ? !sortDesc : false

    this.setState({
      sortCol: colId,
      sortDesc: newSortDesc
    })

    onSort && onSort(activePage, colId, newSortDesc)
  }

  handleSelect(rowId){
    const { selectedRows } = this.state
    const { onRowSelect } = this.props

    const newSelectedRows = selectedRows.has(rowId) ? selectedRows.delete(rowId) : selectedRows.add(rowId)
    this.setState({
      selectedRows: newSelectedRows
    })

    onRowSelect && onRowSelect(rowId, newSelectedRows.toJS())
  }

  handleSelectAll(rowIds){
    const { selectedRows } = this.state
    const { onAllRowsSelect } = this.props

    const newSelectedRows = selectedRows.intersect(rowIds).size === rowIds.length ? selectedRows.subtract(rowIds) : selectedRows.union(rowIds)
    this.setState({
      selectedRows: newSelectedRows
    })

    onAllRowsSelect && onAllRowsSelect(newSelectedRows.toJS())
  }

  render(){
    const { containerClass,
            columns,
            rows,
            children } = this.props

    const { activePage,
            sortCol,
            sortDesc,
            selectedRows } = this.state

    return (
      <div className={containerClass}>
        {Children.map(children, child => {
            if (child.type.name === 'Table')
              return cloneElement(child, {
                ...child.props,
                columns,
                rows,
                sortCol,
                sortDesc,
                selected: selectedRows.toJS(),
                onSort: this.handleSort.bind(this),
                onRowClick: this.handleSelect.bind(this),
                onSelectAll: this.handleSelectAll.bind(this)
              })
            if (child.type.name === 'Paginator')
              return cloneElement(child, {
                 ...child.props,
                 activePage,
                 onPageClick: this.handlePageClick.bind(this)
               })
            return cloneElement(child, {
              activePage,
              sortCol,
              sortDesc,
              selectedRows: selectedRows.toJS(),
              onSort: this.handleSort.bind(this),
              onRowClick: this.handleSelect.bind(this),
              onSelectAll: this.handleSelectAll.bind(this),
              ...child.props
            })
        })}
      </div>
    )
  }
}

export { TableContainer, Table, Paginator }

demo/demo.js
import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import { TableContainer, Table, Paginator } from '../src/TableContainer';

export const TopBar = ({
  rows,
  query,
  onQueryChange,
  onRowDelete,
  sortCol,
  sortDesc,
  selectedRows,
  activePage
}) => {

  return (
    <div className="row" style={{padding: "10px 0"}}>
        <div className="col-sm-6">
            <input
                className="form-control"
                type="text"
                value={query}
                onChange={(e) => onQueryChange(e.target.value, activePage, sortCol, sortDesc)}
                placeholder="Search ..."/>
        </div>
        <div className="col-sm-6 text-right">
            <button
                className="btn btn-danger"
                onClick={onRowDelete.bind(this, selectedRows, activePage)}>
                Delete
            </button>
        </div>
    </div>
  )
}

export const PaginationInfo = ({
    total,
    pageSize,
    activePage
}) => {
    const totalPages = total % pageSize === 0 ? parseInt(total / pageSize) : parseInt(total / pageSize) + 1
    return (
        <div className="pull-left" style={{margin: "20px 0", fontSize: "16px"}}>
            <b>Total rows:</b> <i>{total}</i>
            <span style={{marginLeft: "20px"}}><b>Page:</b> <i>{activePage + 1} / {totalPages}</i></span>
        </div>
    )
}

export const PageSizeSelector = ({
    activePage,
    pageSizeList,
    currentPageSize,
    onChange
}) => {
    return (
        <div className="pull-right btn-group" style={{margin: "20px 20px", fontSize: "16px"}}>
            {pageSizeList.map(size => {
                return(
                    <button
                        className={`btn btn-${currentPageSize===size ? 'success' : 'default'}`}
                        onClick={onChange.bind(this, size, activePage)}>
                            {`${size} per page`}
                     </button>
                )
            })
        }
        </div>
    )
}

class TableDemo extends Component {
    constructor(props){
        super(props)

        const rows = [
            {id: "1", name: "David", gender: "Male", age: "17", height: "176"},
            {id: "2", name: "Kelly", gender: "Female", age: "19", height: "180"},
            {id: "3", name: "James", gender: "Male", age: "12", height: "150"},
            {id: "4", name: "Tim", gender: "Male", age: "39", height: "180"},
            {id: "5", name: "Sam", gender: "Male", age: "49", height: "176"},
            {id: "6", name: "Annie", gender: "Female", age: "26", height: "160"},
            {id: "7", name: "Lucy", gender: "Female", age: "19", height: "170"}
        ]

        this.state = {
            rows: rows,
            filteredRows: rows,
            query: '',
            pageSize: 4,
            pageData: rows.slice(0, 4),
            columns: [
                {id: "name", title: "Name", sortable: true},
                {id: "gender", title: "Gender", template: (v) => <b style={{color: v.gender === "Male" ? 'blue' : 'pink'}}>{v.gender}</b>},
                {id: "age", title: "Age", sortable: true},
                {id: "height", title: "Height", sortable: false}
            ],
            currentPage: 0
        }

        this.reloadData.bind(this)
    }

    reloadData(col, desc, query){
        const { rows } = this.state

        let newRows = rows

        if (col)
            newRows = newRows.sort((a, b) => {
                return desc ?
                    (b[col] > a[col] ? -1 : 1) :
                    (b[col] > a[col] ? 1 : -1)
            })

        newRows = newRows.filter(r => r.name.toLowerCase().indexOf(query.toLowerCase()) > -1)

        return newRows
    }

    handleSort(page, col, desc){
        const { query, pageSize } = this.state

        const newRows = this.reloadData(col, desc, query)

        this.setState({
            filteredRows: newRows,
            pageData: newRows.slice(page*pageSize, (page+1)*pageSize)

        })
    }

    handlePageClick(page, col, desc){
        const { rows, pageSize } = this.state

        this.setState({
            pageData: rows.slice(page*pageSize, (page+1)*pageSize)
        })
    }

    handleQueryChange(query, page, col, desc){
        const { pageSize } = this.state
        const newRows = this.reloadData(col, desc, query)

        this.setState({
            query,
            filteredRows: newRows,
            pageData: newRows.slice(0, pageSize)
        })
    }

    handleDelete(selectedRows, page){
        const { rows, pageSize } = this.state
        const newRows = rows.filter(r => selectedRows.indexOf(r.id) === -1)
        let newPage = page
        if ((page+1)*pageSize > newRows.length)
            newPage =  newRows.length % pageSize === 0 ? parseInt(newRows.length / pageSize) - 1 : parseInt(newRows.length / pageSize)

        this.setState({
            rows: newRows,
            filteredRows: newRows,
            pageData: newRows.slice(newPage*pageSize, (newPage+1)*pageSize)
        })
    }

    handlePageSizeChange(pageSize, page){
        const { rows } = this.state

        let newPage = page
        if ((page+1)*pageSize > rows.length)
            newPage =  rows.length % pageSize === 0 ? parseInt(rows.length / pageSize) - 1 : parseInt(rows.length / pageSize)


        this.setState({
            pageData: rows.slice(newPage*pageSize, (newPage+1)*pageSize),
            pageSize
        })
    }

    render(){
        const { rows, columns, filteredRows, pageData, query, pageSize } = this.state
        return (
            <div style={{padding: "30px"}}>
                <h2 className="text-center"> React Custom Table with Bootstrap </h2>
                <TableContainer
                    containerClass="table-responsive"
                    columns={columns}
                    rows={pageData}
                    onSort={this.handleSort.bind(this)}
                    onPageChange={this.handlePageClick.bind(this)}>
                    <TopBar
                        rows={rows}
                        query={query}
                        onRowDelete={this.handleDelete.bind(this)}
                        onQueryChange={this.handleQueryChange.bind(this)} />
                    <Table
                        tableClass="table table-bordered table-hover"
                        sortableClass="sortable"
                        sortDescClass="desc"
                        sortAscClass="asc"
                        selectedRowClass="selected-row"
                        selectable={true}
                        selectColumnTemplate={(v) => <input type='checkbox' checked={v}/>}
                        selectAllTemplate={(v) => <input type='checkbox' checked={v}/>}/>
                    <PaginationInfo
                        total={filteredRows.length}
                        pageSize={pageSize} />

                    <Paginator
                        paginatorClass="pagination pull-right"
                        total={filteredRows.length}
                        firstPageTemplate={() => <a>{"<<"}</a>}
                        prevPageTemplate={() => <a>{"<"}</a>}
                        pageTemplate={(p) => <a>{p}</a>}
                        nextPageTemplate={() => <a>{">"}</a>}
                        lastPageTemplate={() => <a>{">>"}</a>}
                        pageSize={pageSize}/>
                    <PageSizeSelector
                        pageSizeList={[1, 2, 4]}
                        currentPageSize={pageSize}
                        onChange={this.handlePageSizeChange.bind(this)} />
                </TableContainer>
            </div>
        )
    }
}

ReactDOM.render(<TableDemo />, document.getElementById("demo"));

